GPflow/conditionals.py

```diff
@@ -2,219 +2,106 @@
 from tf_hacks import eye
 import tensorflow as tf
 
-"""
-This file contains functions which compute tensorflow expressions for GP prediction
-equations. This enables much code re-use, because these expressions appear
-frequently. 
-"""
-
-def gp_predict(Xnew, X, kern, F):
-    """
-    Given F, representing the GP at the points X, produce the mean and variance
-    of the GP at the points Xnew.
-
-    We assume K independent GPs, represented by the columns of F. This function
-    computes the Gaussian conditional
-
-        p(F* | F) 
-
-    Xnew is a data matrix, size N x D
-    X are inducing points, size M x D
-    F are function values , size M x K
-
-    See also:
-        gp_predict_whitened -- where F is rotated into V (F = LV)
-        gaussian_gp_predict -- similar, but with uncertainty in F
-
-    """
- 
-    #compute kernel stuff
-    num_data = tf.shape(X)[0]
-    Kdiag = kern.Kdiag(Xnew)
-    Kmn = kern.K(X, Xnew)
-    Kmm = kern.K(X) + eye(num_data)*1e-6
-    Lm = tf.cholesky(Kmm)
-
-    #this is O(N M^2)
-    A = tf.matrix_triangular_solve(Lm, Kmn, lower=True)
-    B = tf.matrix_triangular_solve(tf.transpose(Lm), A, lower=False) # B is Kmm^{-1} Kmn
-
-    #construct the mean and variance of q(f*)
-    fmean = tf.matmul(tf.transpose(B), F)
-    fvar = Kdiag - tf.reduce_sum(tf.square(A), 0)
-    fvar = tf.expand_dims(fvar, 1)
-
-    return fmean, fvar
-
-
-def gaussian_gp_predict(Xnew, X, kern, q_mu, q_sqrt, num_columns):
-    """
-    Given an (approximate) posterior (via q_mu, q_sqrt) to the GP at the points
-    X, produce the mean and variance of the GP at the points Xnew.
-
-    We assume K independent GPs, represented by the columns of q_mu (and the
-    last ax of q_sqrt).  q_mu and q_sqrt are variational posteriors for f, So
-        q(f[:,i]) = N (q_mu[:,i],  diag(q_sqrt[:,i]**2))
-    or
-        q(f[:,i]) = N (q_mu,  W W^T)
-    where W is the lower triangle of q_sqrt[:,:,i]. 
-
-    This function computes the Gaussian integral
-        q(f*) = \int p(f*|f)q(f) df.
-
-    Xnew is a data matrix, size N x D
-    X are inducing points, size M x D
-    q_mu are variational means, size M x K
-    q_sqrt are variational standard-deviations or Cholesky matrices,, size M x K or M x M x K
-    num_columns is the number of columns in q_mu. 
-
-    Note (and TODO):
-        At the moment, num_columns only gets used for the q_sqrt.ndim==3 case,
-        and it tells use the value of q_sqrt.shape()[2]. We need to find a way
-        to get this from the tf graph. 
-
-    See also:
-        gp_predict -- where there is no uncertainty in F
-        gaussian_gp_predict_whitened -- the same, but with whitening (centering) the F variables
-
+def conditional(Xnew, X, kern, f, num_columns, full_cov=False, q_sqrt=None, whiten=False):
```

Comment:
Yes, you're right, it's redundant. The num_columns integer must be the same at the number of columns of f, and the last dimension of q_sqrt. 

The reason its being passed explicitly is so that we can run a python loop over those columns: we need the integer at "python run time", not just in the graph. 

The master branch of theano now has a `tf.scan` function, which we could use for this simple loop, enabling us to remove `num_columns`. That would be a tidy addition in a future version.